### PR TITLE
Feature: Feature cnc leveling in Luban (Experiment, DON'T MERGE)

### PR DESCRIPTION
**Pull request by meiliang1123**

### CODE REVIEW

File: src/server/tasks/LevelingGcode.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        const getOffset = getOffsetCreator(zValues, gridNum, rect);
+        let lastX = 0, lastY = 0, lastZ = 0;
+        let readSize = 0;
+        let lastProgress = 0;
+
+        source.on('line', async (gcodeStr) => {
+            readSize += gcodeStr.length;
+            const progress = Math.floor(readSize / sourceSize * 100);
+            if (progress > lastProgress + 4) {
+                onProgress({ progress, tips: '处理文件中' });
```
Comment (whimsycwd) [inline src/server/tasks/LevelingGcode.js:32]: Please translate to english.

File: src/server/tasks/LevelingGcode.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            if (gcodeStr[0] === 'G' || gcodeStr[0] === 'M') {
+                const gcodeObj = await promiseInterpret(gcodeStr);
+                const x = gcodeObj.X || lastX;
+                const y = gcodeObj.Y || lastY;
+                const z = gcodeObj.Z;
+                if (z === undefined) {
+                    convertedStr = `${gcodeStr} Z${(lastZ + getOffset([x, y, z])).toFixed(3)} `;
+                } else {
+                    if (z % 1 === 0) {
+                        gcodeStr = gcodeStr.replace(`Z${z}.000`, `Z${z}`);
```
Comment (whimsycwd) [inline src/server/tasks/LevelingGcode.js:47]: What happen to other cases? For example. xxx.120, etc. 

Need to fix.

File: src/server/tasks/LevelingGcode.js (hunk 1 above)
Comment (meiliang1123) [reply src/server/tasks/LevelingGcode.js]: done

File: src/server/tasks/LevelingGcode.js (hunk 2 above)
Comment (meiliang1123) [reply src/server/tasks/LevelingGcode.js:47]: no problem with other cases,  numbers like xxx.120 will be replaced  in line 49

Comment (gysnst2) [reply src/server/tasks/LevelingGcode.js:47]: Does this work and with the get be updated with instructions.  Leveling in cnc with the snap maker is needed bad.

**Pull request closed without merging**